### PR TITLE
[vcpkg] Improving android support

### PR DESCRIPTION
**Pull request by christophe-calmejane**
**Describe the pull request**
Improving Android support:
* Allow for custom triplet to be used by the android toolchain: Currently there is a strict name check (eg. arm64-android), which I replaced with a regex match so one can define arm64-android-custom triplet.
* Improved android NDK autodetection algorithm by also checking Xamarin default installed folder
* Fixed boost-modular-build-helper for Android: [improved existing issue which only had a partial fix](https://github.com/microsoft/vcpkg/issues/5809)

**Which triplets are supported/not supported? Have you updated the CI baseline?**
Only affects android toolchain. No change to the CI.

**Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?**
It should.

### CODE REVIEW

**Review by PhoebeHui — COMMENTED** [review]
@christophe-calmejane, thanks for the PR!

Could you update the 'Port-Version' to 3 in boost-modular-build-helper CONTROL file?

**Comment by christophe-calmejane** [comment]
Oh yes sure, forgot I pushed boost fix along with toolchain fix.

**Review by PhoebeHui — APPROVED** [review]
Thanks for your updates!

File: scripts/toolchains/android.cmake [hunk 1]
```diff
@@ -5,13 +5,13 @@ set(ANDROID_TOOLCHAIN clang CACHE STRING "")
 set(ANDROID_NATIVE_API_LEVEL 21 CACHE STRING "")
 set(CMAKE_ANDROID_NDK_TOOLCHAIN_VERSION clang CACHE STRING "")
 
-if (VCPKG_TARGET_TRIPLET STREQUAL "arm64-android")
+if (VCPKG_TARGET_TRIPLET MATCHES "arm64-android")
```
Comment (strega-nil) [inline scripts/toolchains/android.cmake:8]:
```suggestion
if (VCPKG_TARGET_TRIPLET MATCHES "^arm64-android")
```

File: scripts/toolchains/android.cmake [hunk 2]
```diff
@@ -5,13 +5,13 @@ set(ANDROID_TOOLCHAIN clang CACHE STRING "")
 set(ANDROID_NATIVE_API_LEVEL 21 CACHE STRING "")
 set(CMAKE_ANDROID_NDK_TOOLCHAIN_VERSION clang CACHE STRING "")
 
-if (VCPKG_TARGET_TRIPLET STREQUAL "arm64-android")
+if (VCPKG_TARGET_TRIPLET MATCHES "arm64-android")
     set(ANDROID_ABI arm64-v8a CACHE STRING "")
-elseif(VCPKG_TARGET_TRIPLET STREQUAL "arm-android")
+elseif(VCPKG_TARGET_TRIPLET MATCHES "arm-android")
```
Comment (strega-nil) [inline scripts/toolchains/android.cmake:10]:
```suggestion
elseif(VCPKG_TARGET_TRIPLET MATCHES "^arm-android")
```

File: scripts/toolchains/android.cmake [hunk 3]
```diff
@@ -5,13 +5,13 @@ set(ANDROID_TOOLCHAIN clang CACHE STRING "")
...
 set(CMAKE_ANDROID_NDK_TOOLCHAIN_VERSION clang CACHE STRING "")
 
-if (VCPKG_TARGET_TRIPLET STREQUAL "arm64-android")
+if (VCPKG_TARGET_TRIPLET MATCHES "arm64-android")
     set(ANDROID_ABI arm64-v8a CACHE STRING "")
-elseif(VCPKG_TARGET_TRIPLET STREQUAL "arm-android")
+elseif(VCPKG_TARGET_TRIPLET MATCHES "arm-android")
     set(ANDROID_ABI armeabi-v7a CACHE STRING "")
-elseif(VCPKG_TARGET_TRIPLET STREQUAL "x64-android")
+elseif(VCPKG_TARGET_TRIPLET MATCHES "x64-android")
```
Comment (strega-nil) [inline scripts/toolchains/android.cmake:12]:
```suggestion
elseif(VCPKG_TARGET_TRIPLET MATCHES "^x64-android")
```

File: scripts/toolchains/android.cmake [hunk 4]
```diff
@@ -5,13 +5,13 @@ set(ANDROID_TOOLCHAIN clang CACHE STRING "")
...
     set(ANDROID_ABI x86_64 CACHE STRING "")
-elseif(VCPKG_TARGET_TRIPLET STREQUAL "x86-android")
+elseif(VCPKG_TARGET_TRIPLET MATCHES "x86-android")
```
Comment (strega-nil) [inline scripts/toolchains/android.cmake:14]:
```suggestion
elseif(VCPKG_TARGET_TRIPLET MATCHES "^x86-android")
```

**Review by strega-nil — APPROVED** [review]
I'd prefer to have the `^` at the front, just to make sure that we don't match `[something]x64-android` by accident.

File: scripts/toolchains/android.cmake (hunk 1 above)
Comment (christophe-calmejane) [reply scripts/toolchains/android.cmake:8]: Well, that’s the whole point of the change, to allow custom names yet being able to use the android toolchain. 

What toolchain should be used for a triplet named static_x64_android?
Or there are strict rules defined somewhere for triplet naming syntax, even for community ones?

Nevertheless I personally don’t mind as my use case defines a postfix not a prefix, but I’m probably not alone :)

Comment (strega-nil) [reply scripts/toolchains/android.cmake:8]: Triplets shouldn't be named `static_x64_android` -- the method of defining a triplet is `<arch>-<platform>-<other-info>`

Comment (christophe-calmejane) [reply scripts/toolchains/android.cmake:8]: Nice didn't know that, thanks.

**Pull request merged**